lib/off_broadway_twitter/producer.ex

```diff
@@ -3,98 +3,148 @@ defmodule OffBroadwayTwitter.Producer do
+    {:noreply, [], %{state | connection_timer: nil}}
   end
 
   @impl true
-  def handle_demand(demand, state) do
-    IO.inspect(state.size, label: "size")
-    IO.inspect(state.demand + demand, label: "demand")
-
-    handle_received_messages(%{state | demand: state.demand + demand})
+  def handle_info(:connect_to_stream, state) do
+    {:ok, conn} = HTTP2.connect(:https, state.uri.host, state.uri.port)
+
+    {:ok, conn, request_ref} =
+      HTTP2.request(
+        conn,
+        "GET",
+        state.uri.path,
+        [{"Authorization", "Bearer #{state.token}"}],
+        nil
+      )
+
+    {:noreply, [], %{state | request_ref: request_ref, conn: conn, connection_timer: nil}}
   end
 
-  defp handle_received_messages(state) do
-    amount_to_fetch =
-      if state.size >= state.demand do
-        state.demand
-      else
-        0
-      end
+  @impl true
+  def handle_info({:process_stream, %HTTP2{} = conn}, state) do
+    case HTTP2.stream(conn, state.last_message) do
+      {:ok, conn, resp} ->
+        process_responses(resp, %{state | conn: conn})
 
-    {tweets, new_queue} = get_tweets(state.queue, amount_to_fetch, [])
+      {:error, conn, %Mint.HTTPError{reason: {:server_closed_connection, :refused_stream, _}}, _} ->
+        # Do we need to close on our side?
+        HTTP2.close(conn)
 
-    {:noreply, tweets,
-     %{
-       state
-       | queue: new_queue,
-         size: state.size - amount_to_fetch,
-         demand: state.demand - amount_to_fetch
-     }}
-  end
+        timer = schedule_connection(@reconnect_in_ms)
+
+        {:noreply, [], %{state | conn: conn, connection_timer: timer}}
 
-  defp get_tweets(tweets_queue, 0, tweets), do: {Enum.reverse(tweets), tweets_queue}
+      {:error, conn, %{reason: reason}, _} when reason in [:closed, :einval] ->
+        # I got :einval once, but I suppose I was doing something wrong
+        timer = schedule_connection(@connect_in_ms)
 
-  defp get_tweets(tweets_queue, demand, tweets) do
-    {{:value, tweet}, queue} = :queue.out(tweets_queue)
+        {:noreply, [], %{state | conn: conn, connection_timer: timer}}
 
-    get_tweets(queue, demand - 1, [tweet | tweets])
+      {:error, _other} ->
+        {:stop, :stream_stopped_due_unknown_error, state}
+    end
   end
 
   @impl true
-  def handle_info({:data, tweet}, state) do
-    case Jason.decode(tweet) do
-      {:ok, decoded} ->
-        data = Map.fetch!(decoded, "data")
-        meta = Map.delete(data, "text")
-        text = Map.fetch!(data, "text")
-
-        message = %Message{
-          data: text,
-          metadata: meta,
-          acknowledger: {Broadway.NoopAcknowledger, nil, nil}
-        }
-
-        handle_received_messages(%{
-          state
-          | queue: :queue.in(message, state.queue),
-            size: state.size + 1
-        })
+  def handle_info({tag, _socket, _data} = message, state) when tag in [:tcp, :ssl] do
+    send(self(), {:process_stream, state.conn})
```

Comment:
Hrm, why do you send a message to yourself instead of processing the message immediately here? 🤔

Reply:
Good point! :D I misunderstood the cycle of messages thinking that the first message would be sent after the first "stream" call.
My previous implementation was mostly based on that.